### ISSUE TITLE
data structure

### ISSUE THREAD

**Issue by vkonan** [pynapple-org/pynapple#419]
Hi there,

I need some advice. I'm not sure whether this question is supposed to go in the Nemos discussion or here.

I have deconvoluted data from CaImAn of calcium imaging components in the following format:

A list of numpy arrays, each array containing data from one animal, in the shape of <ROI x odors x time (in frames)>.

for example:
len(data_wt) = 8
data[0].shape = (125, 33, 85)

len(data_ko) = 7
data[0].shape = (176, 33, 85)

I'm not sure how i need to structure my data in pynapple so that i can run a classifier in Nemos. I tried creating a TsdTensor (see below), but I'm not sure if I'm reshaping it correctly and I can only make one tensor for one session at a time. Also, I don't know how to contain all the data for both conditions (wt/ko) in one variable so that Nemos can run a classifier.

`tsd_wt = nap.TsdTensor(t = t_array, d = np.reshape(spike_trains_wt[0], (85, 33, 125)))`

Any insight here would be helpful.

Thank you,
Vaibhav

**Comment by gviejo** [comment]
Hi I think the way you created the TsdTensor is correct. From there it depends on what you want to do with nemos. There is an example in the nemos doc for calcium imaging : https://nemos.readthedocs.io/en/latest/tutorials/plot_06_calcium_imaging.html

An alternative for reshaping that does not depend on the actual size is this one
```
np.moveaxis(spike_trains_wt[0], -1, 0)
```
It will move the time axis to the first axis.

**Comment by vkonan** [comment]
Thank you, Guillaume!

The link you referenced uses direct calcium data rather than the deconvolution that we used during the workshop. That is another avenue that I will explore. The current data I am using has been deconvolved. I'm not saying that will cause data handling to be different, but just wanted to note that.

Given that the tensor that I created is data from one animal, how would I handle organizing multiple animals to train a single model?

**Comment by billbrod** [comment]
Hey Vaibhav!

But I just want to point out that nemos doesn't have any classifiers: right now we only have the GLM. [scikit-learn](https://scikit-learn.org/stable/) has classifiers. (We are not planning on adding classifiers)

In either case (for both nemos and scikit-learn), your data needs to be 2d, with time on the first axis, and everything else (ROI, odors, animal) concatenated on the second axis. You could also fit animals independently, depending on what you're trying to do.

What do you want to predict? Typically, GLMs are used to predict some measure of neural activity. In the notebook Guillaume linked, we predict the calcium imaging trace of a single neuron using the animal's head direction and the traces of the other recorded neurons.

From your post, it sounds like you might be trying to use the deconvolved calcium imaging to run a classifier to predict which of two conditions you are in? That's not something that you can do in nemos, that's a scikit-learn problem. 

(I do not think the fact that you're using deconvolved data vs. direct calcium data matters as far as nemos / scikit-learn is concerned. The interpretation will differ, but the way of interacting with the libraries is the same.)

If you do want to rearrange the data to get it 2d, you'd probably do something like the following. From what you've posted, it's a bit hard to tell what all your variables and their types / shapes are, but here's my understanding:
- `data_wt` and `data_ko` are lists of numpy arrays. Each array contains the data from one animal, and each list represents a different condition. Why do they have differnet lengths? Did one animal only do one of the conditions? If so, you should probably drop that array, so that both lists contains 7 arrays.
- Each array in a list has the same shape, with 33 odors, 85 time points, and 125 / 176 ROIs for the wt / ko conditions, respectively. They're shape `(ROI, odor, time)`
- I don't know what `spike_trains_wt` is or what it represents, so I'll ignore it.
- I'm assuming `t_array` is the time stamps you have, which is identical for all animals/conditions

Assuming all that's true, you'd do the following:
```python
import pynapple as nap
import numpy as np
# stack across animals and move the time axis to the front:
# so that data_wt is shape (time, ROI, odor, animal)
data_wt = np.stack([np.moveaxis(d, -1, 0) for d in data_wt], -1)
# do the same thing to data_ko
data_ko = np.stack([np.moveaxis(d, -1, 0) for d in data_ko], -1)

# Convert to TsdTensors with the same shape as above
data_wt = nap.TsdTensor(t=t_array, d=data_wt)
data_ko = nap.TsdTensor(t=t_array, d=data_ko)

# concatenate all the non-time axes:
data_wt = data_wt.reshape((data_wt.shape[0], -1))
data_ko = data_ko.reshape((data_ko.shape[0], -1))

# concatenate the two conditions together:
all_data = np.concatenate([data_wt, data_ko], -1)
```

Then you'd end up with one (2d) TsdFrames with time along the first axis and everything else concatenated along the final axis.

You'll also need to keep track of the labels specifying animal, condition, etc. I'd do that by creating label arrays as the same shape of your original `data_wt` and reshape them as above.

**Comment by BalzaniEdoardo** [comment]
I have to notes on this: reshaping doesn't work the way you want, so do not use that. `np.moveaxis` work or `np.transpose` which re-order the tensor. Reshaping is not equivalent, for example

```python
>>> a = np.arange((50)).reshape(10,5)
>>> a.reshape(5,10) - np.moveaxis(a, 1, 0)
array([[  0,  -4,  -8, -12, -16, -20, -24, -28, -32, -36],
       [  9,   5,   1,  -3,  -7, -11, -15, -19, -23, -27],
       [ 18,  14,  10,   6,   2,  -2,  -6, -10, -14, -18],
       [ 27,  23,  19,  15,  11,   7,   3,  -1,  -5,  -9],
       [ 36,  32,  28,  24,  20,  16,  12,   8,   4,   0]])
```

Second comment, adding to a TsdTensor works, but the time will be initialized at each trial,  therefore the object should not interact with other `pynapple`object that may store the absolute time (time from the start of the experiment) instead of time from beginning of the trial.

As @billbrod said, nemos models works with arrays too, but requires a 2D matrix, while basis can take ND-arrays (as long as the time axis is the first axis) and `compute_features` always spits out a 2D array, concatenating all other dimension on the second axis.
 All the operation described by billy can be done with the following syntax in nemos
```python
import nemos as nmo
import numpy as np

# generate some data for demonstation (4 wt and 3 ko with the same ROIs for simplicity)
data_wt = [np.random.randn(125, 33, 85) for _ in range(4)]
data_ko = [np.random.randn(176, 33, 85) for _ in range(3)]


# concatenate as before
data_wt = np.stack([np.moveaxis(d, -1, 0) for d in data_wt], -1)
data_ko = np.stack([np.moveaxis(d, -1, 0) for d in data_ko], -1)

# use identity basis to do the concatenation and dimension book keeping
basis = nmo.basis.IdentityEval(label="wt") + nmo.basis.IdentityEval(label="ko")
all_data= basis.compute_features(data_wt, data_ko)

```

**Comment by vkonan** [comment]
All of this makes sense! Thanks!

Regarding your understanding of my data: most of what you said is correct, however:
-  the reason that each array is a different size is that each animal's 2-photon recording yields a different number of ROIs (the first animals in the lists have 125 and 176 ROIs in wt and ko, respectively). Does each array have to be the same size or can I supply nemos with different sized arrays? If they need to be the same size, would the solution here be to pad the arrays with nans of the max number of ROIs?
- What I want to accomplish with nemos is to use odors as a feature of the ROI activity (either deconvolved or raw calcium signal as in the demo Guillaume referenced) and predict whether the activity is coming from WT or from KO. I understand that nemos isn't a classifier, but since GLM can be used to predict a binomial distribution, can I not use the logit link function and do what I described? If this question is more appropriate for the discussion section in the nemos repo, let me know.

Thanks again.

**Comment by billbrod** [comment]
Okay, then this is definitely a nemos discussion question!  I'm going to go ahead and close this, can you post a question the nemos page summarizing and linking back to this and we'll continue the conversation there?
